includes/reader-activation/sync/class-metadata.php

```diff
@@ -0,0 +1,397 @@
+	 *
+	 * @param string $key      Metadata key to fetch.
+	 * @param array  $metadata Metadata to fetch from.
+	 *
+	 * @return mixed|null Metadata value or null if not found.
+	 */
+	private static function get_key_value( $key, $metadata ) {
+		if ( isset( $metadata[ $key ] ) ) {
+			return $metadata[ $key ];
+		}
+		if ( isset( $metadata[ self::get_key( $key ) ] ) ) {
+			return $metadata[ self::get_key( $key ) ];
+		}
+		return null;
+	}
+
+	/**
+	 * Get the UTM key from a raw or prefixed key.
+	 *
+	 * @param string $key Key to check.
+	 *
+	 * @return string|false Formatted key if it is a UTM key, false otherwise.
+	 */
+	private static function get_utm_key( $key ) {
+		$keys     = [ 'signup_page_utm', 'payment_page_utm' ];
+		$raw_keys = self::get_raw_keys();
+		foreach ( $keys as $utm_key ) {
+			if ( ! in_array( $utm_key, $raw_keys, true ) ) { // Skip if the UTM key is not in the list of fields to sync.
+				continue;
+			}
+			if ( 0 === strpos( $key, $utm_key ) ) {
+				$suffix = str_replace( $utm_key . '_', '', $key );
+				return self::get_key( $utm_key ) . $suffix;
+			}
+			if ( 0 === strpos( $key, self::get_key( $utm_key ) ) ) {
+				return $key;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Normalizes contact metadata keys before syncing to ESP.
+	 *
+	 * @param array $contact Contact data.
+	 * @return array Normalized contact data.
+	 */
+	public static function normalize_contact_data( $contact ) {
+		if ( ! isset( $contact['metadata'] ) ) {
+			$contact['metadata'] = [];
+		}
+
+		$metadata            = $contact['metadata'];
+		$normalized_metadata = [];
+		$raw_keys            = self::get_raw_keys();
+		$prefixed_keys       = self::get_prefixed_keys();
+
+		// Capture UTM params and signup/payment page URLs as meta for registration or payment.
+		if ( self::has_key( 'current_page_url', $metadata ) || self::has_key( 'payment_page', $metadata ) ) {
+			$is_payment = self::has_key( 'payment_page', $metadata );
+			$raw_url    = false;
+			if ( $is_payment ) {
+				$raw_url = self::get_key_value( 'payment_page', $metadata );
+			} else {
+				$raw_url = self::get_key_value( 'current_page_url', $metadata );
+			}
+
+			$parsed_url = \wp_parse_url( $raw_url );
+
+			// Maybe set UTM meta.
+			if ( ! empty( $parsed_url['query'] ) ) {
+				$utm_key_prefix = $is_payment ? 'payment_page_utm' : 'signup_page_utm';
+				$params         = [];
+				\wp_parse_str( $parsed_url['query'], $params );
+				foreach ( $params as $param => $value ) {
+					$param = \sanitize_text_field( $param );
+					if ( 'utm' === substr( $param, 0, 3 ) ) {
+						$param = str_replace( 'utm_', '', $param );
+						$key   = self::get_key( $utm_key_prefix ) . $param;
+						if ( ! isset( $metadata[ $key ] ) || empty( $metadata[ $key ] ) ) {
+							$metadata[ $key ] = $value;
+						}
+					}
+				}
+			}
+		}
+
+		// Keys allowed to pass through without prefixing.
+		$allowed_keys = [ 'status', 'status_if_new' ];
```

Comment:
Do you think this is a more appropriate place to add the `origin_newspack` key which is currently added in the `upsert` method in [Newsletters](https://github.com/Automattic/newspack-newsletters/pull/1602/files#diff-2c51024c343e46470774b97265cded37f57e841b1db7a53a0f5a6f3e93053224R156)?

Reply:
Good call! I thought the `origin_newspack` was only added when creating a new contact, justifying it not being handled here. It's a good idea to move it.

Reply:
I'm not sure what the purpose of `origin_newspack` is—it doesn't seem like we use it in our code at all. Must be related to automations in the ESP?

Reply:
Here's its birth: https://github.com/Automattic/newspack-newsletters/pull/1091

Maybe @leogermani can share more, but I think it's an identifier so we can treat migration cases or other necessary bulk changes without messing with non-Newspack data from an audience.

Reply:
After some internal discussion it seems we can drop the use of `origin_newspack` entirely as it's no longer needed, but there's nothing needed to do that in this PR.